### ISSUE TITLE
Improve the user experience on Steam Deck

### ISSUE THREAD

**Issue by Ortham** [loot/loot#1893]
There have been several people asking about running LOOT on Linux recently, particularly on the Steam Deck, which (according to Steam's hardware survey) accounts for 42% of all Steam users on Linux (or 0.82% of all Steam users: for comparison, 64-bit Windows 10 & 11 are 95%). LOOT's latest release doesn't run under Wine (see #1886), and we don't currently offer native Linux releases. The Linux builds also have a few limitations compared to the Windows build.

It would be good if we could provide a LOOT user experience on the Steam Deck that's as good as the Windows desktop experience. There are two potential ways to do this:

- Fix the issue with running LOOT's latest release under Wine.
    - This is mostly out of our control, as it's due to a combination of Qt rapidly dropping support for older versions and Wine missing bits of Windows APIs that Qt uses in its newer versions.
    - The latest version of LOOT does run under Wine if it is cross-compiled for Windows from Linux. However, doing so takes a couple of hours and uses gigabytes of storage and so isn't really feasible to do in CI. It's also not clear why it works: e.g. is the functionality that causes crashes in a native Windows build just missing from cross-compiled Qt, and if so is it important?
- Bring Linux builds up to the same level as Windows builds and start releasing them.

I think the second approach is worth investigating.

## Scope

The Steam Deck provides a Linux desktop, so an endless variety of setups is possible, but for my sanity this issue only focuses on getting LOOT working with games that have been bought and installed from Steam. That includes games installed to non-default locations (using Steam's built-in options for having multiple storage locations), but not games installed outside of Steam and then added to Steam, or games managed and played outside of Steam.

## Current Linux limitations

Copy to clipboard functionality depends on xclip being installed, which isn't true by default. I'd like to replace this, the best option is with Qt's `QClipboard` API.

LOOT has the following Windows-specific functionality that doesn't have an equivalent Linux implementation:

- Find games by their Registry keys
- Find games' local app data directories
- Find MS Store games
- Find EGS games

As far as I'm aware, MS Store games aren't available on Linux at all, and while EGS games can be installed through third-party launchers, supporting them is not in scope, so it's OK if finding them doesn't work.

Also, while Valve [say](https://www.steamdeck.com/en/faq) that ext4 with case folding is used for the filesystem of Steam Deck MicroSD cards, according to users on Reddit the internal filesystem uses ext4 without case folding, so there may be issues related to case sensitivity. As these issues may not be obvious, it would be good to detect when a game is installed in a case-sensitive directory and show a warning, but otherwise I don't think it's worth trying to work around: it's better to just install to a case-insensitive directory. That's a bit of a pain if the internal filesystem doesn't support case folding, but as noted in #1095, using ciopfs is one way around that. We'll probably need an FAQ entry about case folding on Linux.

Using the command line, you can check for kernel support using `cat /sys/fs/ext4/features/casefold`, which outputs `supported` in my Manjaro 22.1.3 and Ubuntu 20.04 VMs. You can then check filesystem support using `sudo dumpe2fs -h /dev/sda2 | grep 'Filesystem features' | grep casefold`. If those checks are successful, case folding can be enabled in empty  directories using `chattr +F <path>`, and checked using `lsattr <path>`. However, that may all be specific to ext2/3/4 filesystems, and an easier way for LOOT to check for case-sensitivity is to create two copies of a path that exists, uppercase the filename in one and lowercase the filename in the other, then use `std::filesystem::equivalent()` to see if they still refer to the same item in the filesystem.

## Linux build portability

We do run Linux builds in CI, but discard the artifacts because they're not very portable - they depend on specific system library versions that are unlikely to be present unless running in an environment similar to the build environment (Ubuntu 20.04). We could resolve this by linking LOOT's dependencies statically, but building them as static libraries would be far too slow. The other option is to bundle in the dynamic libraries LOOT needs as part of the build artifact.

There are a few ways to do that, but it seems the best are as an [AppImage](https://appimage.org/) or a [Flatpak](https://flatpak.org/). I can't tell which system is "better" and there's already an issue for publishing LOOT as a Flatpak package (#1757), so that's what I'm trying.

## Steam

Steam installs games under `~/.local/share/Steam/steamapps` by default, e.g. `~/.local/share/Steam/steamapps/common/Oblivion`. It's also possible to install games to other storage locations, which are listed in `~/.local/share/Steam/config/libraryfolders.vdf` and `~/.local/share/Steam/steamapps/libraryfolders.vdf`, which appear to be identical (but they are not symlinked). Those files also list all the app IDs that are present in each library folder. LOOT already uses Steam app IDs during game detection, so they're already known for all supported games.

Each game installed in a library folder has an app manifest file named `appmanifest_<app ID>.acf` that contains information including the name of the game's folder in the `steamapps` folder, so games can be found by reading `libraryfolders.vdf` to find what games' app manifests are installed where, and then reading the app manifests to get the game install directories.

Runtime game files are written to Game files in Steam are written to `<library folder>/steamapps/compatdata/<app ID>/pfx/drive_c`, e.g. `~/.local/share/Steam/steamapps/compatdata/22330/pfx/drive_c/users/steamuser/AppData/Local/Oblivion/Plugins.txt`. If a game is installed to a non-default storage location, its `compatdata` folder is placed there alongside the `steamapps` folder.

All this means that it looks like parsing the VDF and ACF files is a feasible alternative to using the Registry, so LOOT should be able to detect Steam game installs on Linux. On Windows the Steam install path can be found using the Registry entry at `HKEY_LOCAL_MACHINE\Software\WoW6432Node\Valve\Steam\InstallPath`.

### Proton

Proton supplies its own bundled Wine. For Proton Experimental it's at `~/.local/share/Steam/steamapps/common/Proton Experimental/files`. I thought it might be possible to interact with Wine's Registry to find games, but Steam gives each game its own Wine prefix (that's `<library folder>/steamapps/compatdata/<app ID>/pfx` ). If you know where to find that, you're already most of the way to finding the game install, so interacting with the Registry would just involve unnecessary complexity.

Given that, I didn't look into how to interact with the Wine Registry in code, but I could see Fallout 3's Registry keys in regedit when I ran `WINEPREFIX=~/.local/share/Steam/steamapps/compatdata/22370/pfx ~/.local/share/Steam/steamapps/common/Proton Experimental/files/bin/win64 regedit`. However, the installed path is prefixed with a drive letter and uses backslash pretty separators, e.g. `Z:\home\oliver\.local\share\Steam\steamapps\common\Fallout 3 goty\`, so even if LOOT got paths from the Registries, it would have to then adapt them to native Linux paths before using them.

## To Do

- [x] #1757 
- [x] Display a warning message if the current game is installed in an OS / filesystem / directory that does not have case folding enabled.
- [x] Implement the Steam game detection logic described above.
- [x] If user feedback is positive enough to release, update the release CI workflow to include a Linux build that produces an archive and Flatpak. (Following release, we can open a PR to publish on Flathub.)
- [x] Add an FAQ entry with suggestions for how to deal with the case-sensitive filesystem warnings.
- [x] Update https://loot.github.io/docs/help/How-To-Report-Issues-Helpfully.html#supplying-log to include instructions for manually enabling debug logging on Linux (the file path is different).

## Testing

The hard part is setting up a representative test environment. I don't have a Steam Deck, and have had limited success with VMs:

- Valve [suggest](https://partner.steamgames.com/doc/steamdeck/testing) using Manjaro as it's similar to Steam OS. However, in Hyper-V Steam's main UI is invisible and launching Steam hangs after the first launch, so I can't install any games to test with.
- Valve also provide a Steam Deck recovery image, and I followed [this blog post](https://blogs.igalia.com/berto/2022/07/05/running-the-steam-decks-os-in-a-virtual-machine-using-qemu/) to run a Steam Deck VM from a Linux host, but that had the same issues. I then set up GPU passthrough in case the Steam UI invisibility was due to using software rendering, but I think I've not configured the VM to use it correctly as it still seems to be using software rendering even though my Intel integrated GPU is available to it. However, I can see the Steam UI sometimes and was able to install a game - except that it didn't actually download any game files.

Installing Steam directly on my Linux host works fine, which is how I got the information about Steam above, but it means rebooting between Windows and Linux, and it's Ubuntu 22.04, not Manjaro. This means that I don't currently have any way to get personal experience of running LOOT on a Steam Deck, so I'm reliant on others to help test changes.

**Comment by Ortham** [comment]
As of [this build]() I've implemented Steam game detection on Linux and warning messages if the game path or game local path are case-sensitive. I think the warnings would be better if they suggested a solution, it would probably be best to reference the LOOT documentation for that.

In testing, I found that Oblivion creates `Plugins.txt` while LOOT expects `plugins.txt` (which is correct for other games), so case-sensitivity is an easy issue to encounter, and it's not necessarily obvious that it's causing issues.

In my test VM I was able to create a disk image file that used ext4 with its casefold option enabled, and move games into that:

1. Run the following script, changing `IMAGE_LOCATION` and `MOUNT_POINT` to your preference.

```
IMAGE_LOCATION=~/Downloads/steam-icase.img
MOUNT_POINT=~/Downloads/steam-icase

truncate --size=15G "$IMAGE_LOCATION"
mkfs.ext4 -O casefold "$IMAGE_LOCATION"

mkdir -p "$MOUNT_POINT"

sudo mount "$IMAGE_LOCATION" "$MOUNT_POINT"
sudo chown "$USER" "$MOUNT_POINT"

mkdir "$MOUNT_POINT/SteamLibrary"
chattr +F "$MOUNT_POINT/SteamLibrary"
```

2. Now go to Steam's settings -> Storage tab, and add the new mount point at `$MOUNT_POINT` as a storage location.
3. After that, you can move games to that storage location using the options on the same settings page, and then the games' files will be handled case-insensitively.

It's a bit annoying that you can't just enable case-insensitivity in existing filesystems and folders, but I don't think the above is too bad. IIRC it won't persist unless you also update `/etc/fstab` though, and I doubt it's something the average user would be comfortable doing though. It may be better to start by saying that if you're using a Steam Deck, the easiest solution is to move the game to your Deck's microSD card if you have one, as that is case-insensitive by default.

An alternative would be to use https://github.com/adlerosn/cicpoffs, but I'm not sure how reliable that is or if it has significant performance disadvantages for gaming.

**Comment by Ortham** [comment]
I've made a few more improvements as of [this build](https://github.com/loot/loot/actions/runs/5918133292)

- LOOT no longer errors if you try to run it on Linux for a game that you haven't run the launcher for yet (i.e. when the game's local app data folder doesn't exist yet). This should also be the case for users on Windows, unless the game wasn't installed through a launcher.
- The case-sensitive filesystem checks have been improved to work when the local app data folder doesn't exist (LOOT will create it before checking).
- The Flatpak is about half the size it was, as it no longer needs to bundle the ICU libraries.

**Comment by Ortham** [comment]
I've made a few more improvements as of [this build](https://github.com/loot/loot/actions/runs/5980958761):

- LOOT can now detect games installed by a Flatpak version of Steam.
- LOOT can now detect GOG and Epic Games Store games installed using the Heroic Games Launcher.
- LOOT no longer tries to detect sibling game installs on Linux, as it can't detect them correctly.
- LOOT now reads and writes the correct active plugins file for Oblivion (`Plugins.txt` vs. `plugins.txt`), though a case-insensitive filesystem is still recommended as there are a lot of other ways case-sensitivity can cause issues.
- When upgrading its settings, LOOT is now more likely to correctly identify Nehrim/Enderal/Enderal SE.
- Some styling inconsistencies have been fixed.
- The LOOT Flatpak application no longer has general access to the host filesystem, instead it's restricted to default Steam and Heroic Games Launcher install locations, and if you have games installed elsewhere you'll need to grant LOOT access to them.

**Comment by Ortham** [comment]
A tester on Discord just found an issue with the Steam game detection that I've verified on Windows: Steam doesn't reliably add the app ID in libraryfolders.vdf when you install a game or remove it when you uninstall the game. Adding or removing a library location does update the app IDs listed, and when I installed the game in a new library location it did immediately appear. Relaunching Steam does update the file, the app IDs were updated when Steam started.

Given that, I think it's best if LOOT doesn't just look for the listed app's manifest files, but instead speculatively tries to open the manifest file for each supported game.
